### PR TITLE
sjawhar-legion-272: add repetitive tool use circuit breaker

### LINKED ISSUE #272: OMO Replacement: Circuit Breaker for Repetitive Tool Use (T3)
## Context
Part of #200 — Replace oh-my-opencode (OMO) with opencode-legion.

**Wave 1 — Foundation** | Blocks: T22 | Blocked By: None

## Problem Statement
Detect and interrupt agents that call the same tool with identical arguments repeatedly, indicating a stuck loop.

## What to Do
- Implement as `tool.execute.before` or `tool.execute.after` hook
- Track per-session: tool name + args hash
- Configurable threshold (default 5 identical calls in a window)
- When triggered: inject warning message, optionally abort
- Config: threshold count, window size, action (warn/abort)
- Clean up tracking on session.deleted

## Must NOT Do
- Do NOT track across sessions
- No complex similarity matching — exact arg match is sufficient

## Existing Infrastructure
- `src/hooks/` — 12 existing hooks following consistent patterns
- `src/hooks/utils.ts` — shared utilities (isRecord, extractTodos, resolveSessionID)
- `src/index.ts` — hook registration in plugin entry
- Hook points: `tool.execute.before`, `tool.execute.after`, `event` (for session.deleted cleanup)

## Acceptance Criteria
- [ ] TDD: failing test first, then implementation
- [ ] 5 identical tool calls (same tool + same args) triggers the breaker
- [ ] Warning/abort behavior is configurable via plugin config
- [ ] Per-session tracking — different sessions have independent counters
- [ ] Tracking cleaned up on session.deleted event
- [ ] 4 identical calls does NOT trigger (below threshold)

## Testing Infrastructure
- `bun test` available
- Follow patterns from `src/hooks/__tests__/output-compression.test.ts`

## QA
- Simulate 5 identical Read calls → verify trigger
- Simulate 4 identical → verify no trigger
- Simulate 5 across 2 sessions → verify independent tracking
Evidence: `.sisyphus/evidence/task-3-circuit-breaker.txt`

## Commit Message
`feat(legion): add repetitive tool use circuit breaker`

## Skills
None required.

## Routing Hints
- Complexity: quick
- Estimated implementers: 1

### PR DESCRIPTION
**Pull request by sjawhar**
Closes #272

## Summary

Adds a `tool.execute.before` hook that detects when an agent calls the same tool with identical arguments repeatedly (indicating a stuck loop), and either warns or aborts the call when a configurable threshold is reached.

### Changes
- **Config** (`config/index.ts`): Added `CircuitBreakerConfig` interface, Zod schema, and merge function. Fields: `enabled` (boolean), `threshold` (number, default 5), `action` ("warn" | "abort", default "abort").
- **Hook** (`hooks/circuit-breaker.ts`): Factory function returning `tool.execute.before` handler and `event` handler. Tracks `Map<sessionID, Map<"toolName:argsHash", count>>`. Uses stable key-sorted JSON for arg comparison. Cleans up on `session.deleted`.
- **Wiring** (`index.ts`): Circuit breaker runs first in `tool.execute.before` chain. Session cleanup wired into `session.deleted` event block.
- **Tests** (`hooks/__tests__/circuit-breaker.test.ts`): 26 tests covering threshold triggering, per-session isolation, args normalization (key order, nested objects, null/empty), action configuration (abort/warn), session cleanup, disabled state, and edge cases.

### Acceptance Criteria
- [x] TDD: failing test first, then implementation
- [x] 5 identical tool calls (same tool + same args) triggers the breaker
- [x] Warning/abort behavior is configurable via plugin config
- [x] Per-session tracking — different sessions have independent counters
- [x] Tracking cleaned up on session.deleted event
- [x] 4 identical calls does NOT trigger (below threshold)

### CODE REVIEW

**Comment by sjawhar** [comment]
## Behavioral Test Results

### Summary
**PASS** — 6/6 acceptance criteria verified

### Results
| Criterion | Status | Evidence |
|-----------|--------|----------|
| TDD: failing test first, then implementation | ✅ | Implementer attestation in retro comment + PR body. Branch was squash-merged so individual TDD commits not visible in log. |
| 5 identical tool calls (same tool + same args) triggers the breaker | ✅ | `bun test --grep "triggers on the 5th"` — 1 pass, 0 fail. Test creates hook with threshold=5, calls 4 times, then 5th call throws `/repetitive tool use detected/i`. |
| Warning/abort behavior is configurable via plugin config | ✅ | `bun test --grep "throws on trigger when action is"` — 1 pass (abort throws). `bun test --grep "does not throw on trigger when action is"` — 1 pass (warn does not throw). `bun test --grep "logs a warning"` verifies console.warn is called. Config wired at `index.ts:94` via `pluginConfig.circuitBreaker`. |
| Per-session tracking — different sessions have independent counters | ✅ | `bun test --grep "5 calls across 2 sessions"` — 1 pass. 3 calls in s-1 + 2 in s-2 = 5 total but neither triggers. `bun test --grep "triggers independently per session"` — s-1 triggers at 5, s-2 still unaffected at 4 calls. |
| Tracking cleaned up on session.deleted event | ✅ | `bun test --grep "clears tracking on session.deleted"` — 1 pass. After 4 calls + session.deleted, 4 more calls do not trigger (counter reset). Also tested: other sessions unaffected by deletion, unknown session deletion is a no-op, info.id format supported. |
| 4 identical calls does NOT trigger (below threshold) | ✅ | `bun test --grep "does not trigger on 4"` — 1 pass, 4 expect() calls all pass (none throw). |

### Full Suite Results
- **Circuit breaker tests:** 26 pass, 0 fail (75 expect() calls)
- **Full opencode-plugin suite:** 384 pass, 0 fail (872 expect() calls)
- **TypeScript:** Clean (`bunx tsc --noEmit` — exit 0)
- **Biome lint:** Clean (`bunx biome check src/` — 83 files checked, no fixes)

### Test Quality Critique
- **Meaningful, not decorative.** Tests exercise the actual hook function with real inputs — no excessive mocking. The only mock is `spyOn(console, "warn")` for the warn-action test, which is appropriate.
- **Good edge case coverage:** null args, empty args, nested objects with different key order, missing sessionID, `info.id` format for session properties, non-`session.deleted` events.
- **Key design decision tested explicitly:** output.args vs input.args distinction verified in "uses output.args for hashing" test (line 423-442).
- **Minor gap:** No config merge tests for `mergeCircuitBreaker` specifically. The function is trivial (`{ ...base, ...override }`) so risk is low. No integration test verifying the wiring in index.ts — mitigated by TypeScript type checking.
- **Verdict:** Tests are solid and would catch real regressions.

### Documentation Feedback
- No user-facing documentation updated (appropriate — this is an internal plugin hook)
- Code has good JSDoc comments explaining the hook's behavior and tracking structure
- PR body adequately describes the feature, config options, and changes

### Observations
- The `console.warn` output from the warn-path test is visible in test runner output — expected behavior, not a defect
- Config defaults are sensible: enabled=true, threshold=5, action=abort
- The hook runs first in the `tool.execute.before` chain (before `subagentQuestionBlockerHook`), which is correct — a stuck loop should be caught before other validations

### Config Compliance
- require_specific_task: false — not configured
- require_taiga_evidence: false — not configured

**Comment by sjawhar** [comment]
## Review Summary

**CRITICAL (P1):** 0 issues
**IMPORTANT (P2):** 0 issues
**MINOR (P3):** 2 suggestions

**Verdict: Approved to merge.** Clean implementation that meets all acceptance criteria.

---

### Spec Compliance

All 6 acceptance criteria verified:

| Criterion | Status | Evidence |
|-----------|--------|----------|
| TDD: failing test first, then implementation | PASS | Retro attestation; 26 tests written before/alongside impl |
| 5 identical tool calls triggers the breaker | PASS | `count >= threshold` at circuit-breaker.ts:68; test line 25-36 |
| Warning/abort behavior configurable via plugin config | PASS | `CircuitBreakerConfig.action` with Zod schema + merge function |
| Per-session tracking — independent counters | PASS | `Map<sessionID, Map<key, count>>`; test lines 84-122 |
| Tracking cleaned up on session.deleted event | PASS | `sessionCounts.delete(sessionID)`; test lines 285-381 |
| 4 identical calls does NOT trigger | PASS | `count >= threshold` means 4 < 5 = no trigger; test line 14-23 |

### Code Quality

**Strengths:**
- Follows existing hook patterns precisely (factory function returning method-keyed object, same as `createOutputCompressionHook`, `createStopContinuationGuardHook`, etc.)
- Correctly uses `output.args` (parsed tool input) not `input.args` — a known gotcha documented in the retro
- Stable JSON hashing via recursive key-sorted replacer handles nested objects correctly
- Config integration is complete: Zod schema, TS interface, `PluginConfig` field, `PluginConfigSchema` field, `mergeConfig` entry — all 5 touch points
- Uses shared utilities (`isRecord`, `resolveSessionID`) from `./utils.ts` — no local reimplementation
- Hook wired first in `tool.execute.before` chain (before `subagentQuestionBlockerHook`), which is correct — circuit breaker should fire before other checks
- Session cleanup wired inside the `session.deleted` block, consistent with other hooks

**Test quality:**
- 26 tests, 75 expect() calls — thorough coverage
- Tests cover: threshold triggering, per-session isolation, args normalization (key order, nested, null/empty), action config (abort/warn/message content), session cleanup (delete/unknown/cross-session/non-delete events/info.id format), disabled state, edge cases (missing sessionID, output.args precedence)
- Tests exercise the real hook with real inputs — no excessive mocking
- Console.warn spy test properly restores mock in finally block

### Minor Suggestions (P3)

1. **Console.warn leaks in test output** — the "does not throw on trigger when action is 'warn'" test (line 234-244) doesn't spy on `console.warn`, so the warning message appears in test runner output. Cosmetic only — no functional impact.

2. **No reset mechanism after threshold** — once fired, every subsequent call with the same args continues to trigger. The test at line 66-81 explicitly verifies this is intentional behavior. Worth noting in case a future "reset after N more calls" or "cooldown" feature is desired, but not a problem for current requirements.

### Evidence
- CI status: lint PASS, test PASS (384 tests), typecheck PASS
- Local verification: `bun test src/hooks/__tests__/circuit-breaker.test.ts` — 26 pass, 0 fail, 75 expect() calls
- Full suite: `bun test` — 384 pass, 0 fail
- No bot review comments on PR
- No existing reviews on PR

File: packages/opencode-plugin/src/hooks/__tests__/circuit-breaker.test.ts [hunk 1]
```diff
@@ -0,0 +1,444 @@
...
+      const hook = createCircuitBreakerHook({ threshold: 5, action: "warn" });
+      const args = { command: "ls" };
+
+      for (let i = 0; i < 4; i++) {
+        hook["tool.execute.before"](makeInput("bash", "s-1", args), makeOutput(args));
+      }
+
+      expect(() =>
+        hook["tool.execute.before"](makeInput("bash", "s-1", args), makeOutput(args))
+      ).not.toThrow();
```
Comment (sjawhar) [inline packages/opencode-plugin/src/hooks/__tests__/circuit-breaker.test.ts:244]: **P3:** Console.warn leaks to test runner output. The "does not throw on trigger when action is 'warn'" test exercises the warn path without spying on `console.warn`, so the warning message appears in `bun test` output. The spy-based test at line 247 does suppress it — this test just doesn't. Cosmetic only.

File: packages/opencode-plugin/src/hooks/circuit-breaker.ts [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    let sessionMap = sessionCounts.get(sessionID);
+    if (!sessionMap) {
+      sessionMap = new Map<string, number>();
+      sessionCounts.set(sessionID, sessionMap);
+    }
+
+    const count = (sessionMap.get(key) ?? 0) + 1;
+    sessionMap.set(key, count);
+
+    if (count >= threshold) {
```
Comment (sjawhar) [inline packages/opencode-plugin/src/hooks/circuit-breaker.ts:68]: **P3:** Once `count >= threshold`, every subsequent call with the same args continues to trigger (count increments to 6, 7, 8...). This is correct behavior for catching stuck loops — the test at line 66 ("continues to trigger on subsequent calls after threshold") explicitly verifies this. Noting for documentation: there's no built-in reset/cooldown. If a future need arises for "warn N times then stop warning," a reset mechanism would be needed here.
